articles/heap-data-structure-python/index.md

```diff
@@ -0,0 +1,202 @@
+
+For more background on the different types of data structures in Python, check out the following articles:
+
+- [Introduction to Data Structures](/engineering-education/data-structures-python-part-1/)
+- [List](/engineering-education/list-data-structure-python/)
+- [Stack](/engineering-education/stack-data-structure-python/)
+- [Queue](/engineering-education/queue-data-structure-python/)
+- [Linked Lists](/engineering-education/linked-list-data-structure-python/)
+- [Binary Trees](/engineering-education/binary-tree-data-structure-python/)
+
+_Note: Prerequisites -- Make sure you have basic Python knowledge before diving into this article. It also might be a good idea to check out some linear data structures. (links are given above)_
+
+### Table of Contents
+
+- [Heaps: Introduction](#heaps:-introduction)
+- [Applications of Heaps](#applications-of-heaps)
+- [Implementing a Heap](#implementing-a-heap)
+- [Practice Heaps](#practice-heaps)
+- [Conclusion](#conclusion)
+
+### Heaps: Introduction
+
+Heaps are complete binary trees. Complete binary trees satisfy the following conditions:
+
+- All levels are filled, except the last.
+- All the nodes are as far left as possible.
+
+![Complete Binary Tree](/engineering-education/heap-data-structure-python/complete-binary-tree.png)
+
+_Figure: Complete Binary Tree ([Source](https://www.andrew.cmu.edu/course/15-121/lectures/Trees/trees.html))_
+
+Heaps satisfy the heap property. This means that the root of every subtree should be the greatest or smallest element in the subtree, recursively.
+
+### Applications of Heaps
+
+- Priority Queues can be implemented using heaps. The root of a heap always contains the maximum or the minimum value, based on the heap type. Therefore, a min-priority queue is implemented using a minheap. A max-priority queue is implemented using a maxheap. The element with the highest priority can be retrieved in O(1) time.
+
+- Statistics -- If we want to get order statistics, heaps serve as a great choice. If we want the kth smallest or largest element, we can pop the heap k times to retrieve them.
+
+- Heaps are used to implement various graph algorithms like Dijkstra's algorithm and Prim's algorithm.
+
+### Implementing a Heap
+
+#### Heap Operations
+
+A heap has the following methods:
+
+- **`getMax()`**
+
+  - This operation returns the root of the maxheap.
+  - Time Complexity - O(1).
+
+- **`insert(k)`**
+
+  - This operation insert the key **k** into the heap.
+  - Then it rearranges the heap to restore the heap property.
+  - Time Complexity - O(log n).
+
+- **`heapify()`**
+
+  - This operation restored the heap property by rearranging the heap.
+  - Time complexity - O(log n).
+
+- **`printHeap()`**
+
+  - Prints the heap's level order traversal.
+
+#### Maxheap using List
+
+We are going to do the list implementation of a heap. In this, the heap's level-order traversal would be stored as an array/list.
```

Comment:
The picture doesn't do a great job of explaining what Level-Order Traversal is. The Quora page you gave explains the order the heap is traversed in, which would help a lot.

Reply:
> The picture doesn't do a great job of explaining what Level-Order Traversal is. The Quora page you gave explains the order the heap is traversed in, which would help a lot.

I've explained the diagram in detail and have included a link to the Quora page. 😄